elastic/beats
# [aws] [s3] Introduce ignore_older & start_timestamp for S3 input allowing better registry cleanups

## Proposed commit message

Introduce `ignore_older` and `start_timestamp` properties to AWS S3 input. This is a follow-up for [#41694](https://github.com/elastic/beats/pull/41694).

The configurations introduced here act as input object filters. If the object fails to match derived filters, the entries will be cleaned up from the registry, reducing filebeat memory consumption.

Introduced configurations are, 

- ignore_older : Accepts a time duration in which entries are accepted for processing
- start_timestamp: A timestamp from which objects are accepted for processing

For both inputs, the object's last modified timestamp is taken into comparison. See **Use cases** section for further explanation

## Checklist


##  Disruptive User Impact

None as defaults are disabled. However, when configurations introduced here are used, the following can have an impact on the user,

-  When`start_timestamp` is defined, then objects with the last modified timestamps prior to the timestamp are ignored from processing   (documented [^1])
- When `ignore_older` is defined, then objects that do not fall within the look-back period when processing starts (polling run) are ignored  (documented [^1])
- When both `start_timestamp` & `ignore_older` are defined, the initial run will process all entries up to `start_timestamp`. The subsequent runs will not include entries that do not fall within `ignore_older` even if processing failed for an object. (documented [^1])


## How to test this PR locally

- Build filebeat from the changest included in the PR 
- Source S3 bucket with objects (you may use this tool [^2] to create entries)
- Try configuring filebeat with alternative values for `ignore_older` & `start_timestamp` to see how data ingestion change with their values. See **Use cases** section for further explanation
 
## Related issues

- Depends on https://github.com/elastic/beats/pull/41694
- Fixes https://github.com/elastic/beats/issues/39116
- Fixes https://github.com/elastic/beats/issues/41232

## Use cases

Consider below diagrams where there're 3 objects Object A, Object B and Object C with their last modified timestamps of t1, t2 and t3.

And consider how filebeat processes and tracks registry entries based on the following scenarios

### Default behavior 

If none of the configurations are used, then filebeat will process and the internal registry will track all objects continuously unless they are removed from the bucket.

![image](https://github.com/user-attachments/assets/51cc00bd-9ab5-46bd-86cd-36a83a824aca)


### Use start_timestamp

If `start_timestamp` is used, objects newer than the timestamp are accepted for processing. The registry will grow unless objects are removed from the bucket by other means (ex:- lifecycle policy).

![image](https://github.com/user-attachments/assets/c363cd96-5a16-465a-86ec-468a1c3020d2)

### Use ignore_older

If `ignore_older` is defined, input will process objects within the provided duration, calculated from the current time. The registry will track objects within the current timeframe and others will get cleaned up eventually by subsequent runs.

![image](https://github.com/user-attachments/assets/65ea96b6-469c-47f7-b084-635080af5836)

### Use both ignore_older & start_timestamp

If both properties are defined, 

- The initial run will include entries within the start_timestamp (ignoring `ignore_older` duration). 
- Subsequent runs will only consider entries within the `ignore_older` duration. 

![image](https://github.com/user-attachments/assets/1e1f2046-765e-45dc-bc35-2be9b83ea5e4)


[^1]: https://github.com/elastic/beats/pull/41817/files#diff-422765b7341c5bbf6de7af38927e34e00a5073b188585a7af3c4fee1175b64a6
[^2]: https://github.com/Kavindu-Dodan/data-gen
